# 📊 TechCorp Dataset — Data Cleaning & Analysis

## Phase 1 — Load & Explore

In this phase we load all four sheets from the Excel file into separate DataFrames, then get a first look at their shapes, columns, data types, and value distributions.

In [1]:
import pandas as pd
import numpy as np

In [2]:
#Load all 4 sheets in one shot
sheets = pd.read_excel('project_data.xlsx', sheet_name=None)

In [3]:
#show sheets
sheets.keys()

dict_keys(['Employees', 'Departments', 'Products', 'Sales'])

In [4]:
#Unpack all sheets into separate dataframes
employees = sheets['Employees']
dept = sheets['Departments']
prod = sheets['Products']
sales = sheets['Sales']

In [21]:
employees.head()

,employee_id,full_name,email,department,job_title,city,country,hire_date,salary,performance_rating,manager_id
0,1001,Gregory White,gregory.white@techcorp.com,Sales,Account Manager,Mumbai,India,2021-02-06,119704.35,4.9,1006
1,1002,Christine Sanders,christine.sanders@techcorp.com,HR,HR Generalist,Amsterdam,Netherlands,2022-06-10,54396.61,3.6,1003
2,1003,Donald Wright,donald.wright@techcorp.com,HR,Recruiter,Cape Town,South Africa,2018-11-27,55221.15,1.9,1002
3,1004,Richard Edwards,richard.edwards@techcorp.com,Engineering,QA Engineer,São Paulo,Brazil,2018-03-04,119468.82,2.6,1007
4,1005,Christopher Nelson,christopher.nelson@techcorp.com,Finance,Controller,Dubai,UAE,2022-06-29,NaN,3.2,1008


In [22]:
dept.head()

,department_id,department_name,department_head,office_location
0,101,Engineering,Alexandra Stone,San Francisco
1,102,Sales,Marcus Reid,New York
2,103,HR,Priya Sharma,London
3,104,Finance,Daniel Osei,Chicago
4,105,Marketing,Laura Chen,Berlin


### Shapes of each DataFrame

Check how many rows and columns each sheet contains:

- `employees.shape`
- `dept.shape`
- `prod.shape`
- `sales.shape`

In [24]:
print(employees.shape)
print(dept.shape)
print(prod.shape)
print(sales.shape)

(123, 11)
(6, 4)
(35, 5)
(410, 9)


### Column Names

Inspect the column names for each DataFrame:

- `employees.columns`
- `dept.columns`
- `prod.columns`
- `sales.columns`

In [ ]:
employees.columns.to_list()

['sale_id',
 'date',
 'employee_id',
 'product_id',
 'region',
 'units_sold',
 'unit_price',
 'discount',
 'total_revenue']

In [26]:
dept.columns.to_list()

['department_id', 'department_name', 'department_head', 'office_location']

In [27]:
prod.columns.to_list()

['product_id', 'product_name', 'category', 'price', 'launch_date']

In [28]:
sales.columns.to_list()

['sale_id',
 'date',
 'employee_id',
 'product_id',
 'region',
 'units_sold',
 'unit_price',
 'discount',
 'total_revenue']

### Data Types

Use `.dtypes` to identify columns that may need type conversion (e.g. date strings stored as `object`).

In [29]:
print(sales.dtypes)

sale_id            int64
date              object
employee_id        int64
product_id         int64
region            object
units_sold         int64
unit_price       float64
discount         float64
total_revenue    float64
dtype: object


### Full DataFrame Info

`.info()` gives a concise summary — column names, non-null counts, and dtypes — all in one view.
Uncomment the DataFrame you want to inspect:

In [ ]:
# employees.info()
# dept.info()
# prod.info()
# sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   employee_id         123 non-null    int64  
 1   full_name           123 non-null    object 
 2   email               123 non-null    object 
 3   department          123 non-null    object 
 4   job_title           123 non-null    object 
 5   city                123 non-null    object 
 6   country             123 non-null    object 
 7   hire_date           123 non-null    object 
 8   salary              118 non-null    float64
 9   performance_rating  118 non-null    float64
 10  manager_id          123 non-null    int64  
dtypes: float64(2), int64(2), object(7)
memory usage: 10.7+ KB


### Statistical Summary

`.describe()` returns count, mean, std, min/max, and quartiles for all numeric columns.
Uncomment the DataFrame you want to inspect:

In [34]:
# employees.describe()
# dept.describe()
# prod.describe()
# sales.describe()

### Unique Values in Key Columns

Check the number of unique IDs to spot potential duplicates early.

In [36]:
print(len(employees['employee_id'].unique()))

120


In [13]:
employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   employee_id         123 non-null    int64  
 1   full_name           123 non-null    object 
 2   email               123 non-null    object 
 3   department          123 non-null    object 
 4   job_title           123 non-null    object 
 5   city                123 non-null    object 
 6   country             123 non-null    object 
 7   hire_date           123 non-null    object 
 8   salary              118 non-null    float64
 9   performance_rating  118 non-null    float64
 10  manager_id          123 non-null    int64  
dtypes: float64(2), int64(2), object(7)
memory usage: 10.7+ KB


### Value Frequency per Categorical Column

`.value_counts()` shows how often each category appears — useful for spotting imbalances or typos.

In [37]:
employees['department'].value_counts()

department
Engineering         26
Sales               25
Customer Support    19
Finance             19
HR                  17
Marketing           17
Name: count, dtype: int64

In [38]:
sales['region'].value_counts()

region
Asia             94
Africa           88
Europe           82
North America    74
Middle East      72
Name: count, dtype: int64

In [42]:
prod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    35 non-null     int64  
 1   product_name  35 non-null     object 
 2   category      35 non-null     object 
 3   price         32 non-null     float64
 4   launch_date   35 non-null     object 
dtypes: float64(1), int64(1), object(3)
memory usage: 1.5+ KB


In [45]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 410 entries, 0 to 409
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   sale_id        410 non-null    int64  
 1   date           410 non-null    object 
 2   employee_id    410 non-null    int64  
 3   product_id     410 non-null    int64  
 4   region         410 non-null    object 
 5   units_sold     410 non-null    int64  
 6   unit_price     410 non-null    float64
 7   discount       405 non-null    float64
 8   total_revenue  410 non-null    float64
dtypes: float64(3), int64(4), object(2)
memory usage: 29.0+ KB


### Issue Log — TechCorp Dataset (Phase 1 Findings)

| Sheet | Issue | Detail |
|-------|-------|--------|
| **Employees** | Extra rows | 123 rows found, expected 120 → likely 3 duplicate rows |
| **Employees** | Missing values | `salary`: 5 missing; `performance_rating`: 5 missing |
| **Employees** | Wrong dtype | `hire_date` stored as `object` — needs datetime conversion |
| **Employees** | Name casing | Some names in lowercase / UPPERCASE → needs standardisation |
| **Products** | Missing values | `price`: 3 missing values |
| **Sales** | Extra rows | 410 rows found, expected 400 → likely ~10 duplicate rows |
| **Sales** | Missing values | `discount`: 5 missing values |
| **Sales** | Calculation error | Some `total_revenue` values may be incorrect |
| **Sales** | Wrong dtype | `date` stored as `object` — needs datetime conversion |
| **Departments** | — | ✅ No issues found — clean reference table |

---

## Phase 2 — Data Selection & Filtering

Practise selecting columns and filtering rows using boolean conditions on our TechCorp dataset.

In [5]:
# Exercise 1: Simple column selection
# Q: list of employees with just their names, department,job title, salary
employees[['full_name', 'department', 'job_title', 'salary']].head()

,full_name,department,job_title,salary
0,Gregory White,Sales,Account Manager,119704.35
1,Christine Sanders,HR,HR Generalist,54396.61
2,Donald Wright,HR,Recruiter,55221.15
3,Richard Edwards,Engineering,QA Engineer,119468.82
4,Christopher Nelson,Finance,Controller,NaN


In [6]:
# Exercise 2: Single condition 
# Q: show all employees from london office
employees[employees['city'] == 'London'].head()

,employee_id,full_name,email,department,job_title,city,country,hire_date,salary,performance_rating,manager_id
55,1056,Joshua Brown,joshua.brown@techcorp.com,Engineering,Data Engineer,London,UK,2024-07-06,145878.77,2.1,1007
91,1092,Donald Thompson,donald.thompson@techcorp.com,Marketing,Marketing Specialist,London,UK,2018-01-11,84936.93,3.8,1025
114,1115,Carol Anderson,carol.anderson@techcorp.com,HR,Talent Acquisition Specialist,London,UK,2023-02-02,40356.98,3.9,1002
118,1119,Nicholas Parker,nicholas.parker@techcorp.com,Marketing,Digital Marketing Manager,London,UK,2023-04-29,89507.67,NaN,1025


In [9]:
# Exercise 3: Multiple conditions
# # Q: show all employees from london office in the Marketing department

employees[(employees['city'] == 'London') & (employees['department'] == 'Marketing')].head()


,employee_id,full_name,email,department,job_title,city,country,hire_date,salary,performance_rating,manager_id
91,1092,Donald Thompson,donald.thompson@techcorp.com,Marketing,Marketing Specialist,London,UK,2018-01-11,84936.93,3.8,1025
118,1119,Nicholas Parker,nicholas.parker@techcorp.com,Marketing,Digital Marketing Manager,London,UK,2023-04-29,89507.67,NaN,1025


In [10]:
# Exercise 3: Multiple conditions OR
# Q: show all employees from the customer-facing team - Sales and customer services
employees[(employees['department'] == 'Sales') | (employees['department'] == 'Customer Support')].head()

,employee_id,full_name,email,department,job_title,city,country,hire_date,salary,performance_rating,manager_id
0,1001,Gregory White,gregory.white@techcorp.com,Sales,Account Manager,Mumbai,India,2021-02-06,119704.35,4.9,1006
5,1006,Joseph Kelly,joseph.kelly@techcorp.com,Sales,Business Development Rep,Amsterdam,Netherlands,2022-02-19,NaN,3.4,1001
8,1009,Andrew Parker,andrew.parker@techcorp.com,Customer Support,Data Analyst,Mumbai,India,2019-09-27,60888.77,1.9,1012
11,1012,Helen Cook,helen.cook@techcorp.com,Customer Support,Support Lead,San Francisco,USA,2020-04-13,69376.19,1.5,1009
22,1023,Deborah Howard,deborah.howard@techcorp.com,Sales,Sales Manager,San Francisco,USA,2022-11-22,40186.04,4.7,1001


### ⚠️ Common Filtering Mistakes

| # | Mistake | Fix |
|---|---------|-----|
| 1 | No parentheses around individual conditions | Wrap each condition in `()` |
| 2 | Single `[]` for multiple columns | Use double brackets `[[col1, col2]]` |
| 3 | Using `and` instead of `&` | Use `&` for AND between conditions |
| 4 | Using `or` instead of `\|` | Use `\|` for OR between conditions |
| 5 | Using `=` instead of `==` | Use `==` for equality checks |
| 6 | Forgetting to store filtered output | Assign result to a new variable before further operations |

---

## Phase 3 — Finding & Fixing Missing Values

> **Golden rules:**
> - You cannot *ignore* missing values — they will skew statistics and break models.
> - You cannot fill them with made-up values without a sound strategy.

**Three standard strategies:**
1. **Drop** rows/columns with missing values (`dropna()`)
2. **Fill** with a fixed value — e.g. `0` or `'Unknown'`
3. **Impute** with a statistic — mean, median, or mode (often grouped by category)

In [2]:
import pandas as pd
import numpy as np

sheets = pd.read_excel('project_data.xlsx', sheet_name=None)
#Unpack all sheets into separate dataframes
employees = sheets['Employees']
dept = sheets['Departments']
prod = sheets['Products']
sales = sheets['Sales']

### Step 1 — Detect All Missing Values

Loop over every sheet and print columns that have at least one `NaN`.

In [3]:
for name, df in sheets.items():
    print(f"\n {'='*40}")
    print(f"    Sheet: {name}")
    print(f"{'='*40}")
    missing = df.isna().sum()
    print(missing[missing > 0] if missing.any() else "No missing values") 


    Sheet: Employees
salary                5
performance_rating    5
dtype: int64

    Sheet: Departments
No missing values

    Sheet: Products
price    3
dtype: int64

    Sheet: Sales
discount    5
dtype: int64


### Step 2 — Missing Values as a Percentage

A helper function that shows the **count** and **percentage** of missing values per column,
making it easier to decide which strategy to apply.

In [12]:
def missing_summary(df, name):
    total = len(df)
    missing = df.isna().sum()
    missing = missing[missing > 0]
    if len(missing) == 0:
        print(f"\n{name}: ✅ No missing values")
        return
    pct = (missing / total * 100).round(2)
    summary = pd.DataFrame({'missing_count': missing, 'missing_%': pct})
    print(f"\n{name}:")
    print(summary)

missing_summary(employees,   "Employees")
# missing_summary(prod,  "Products")
# missing_summary(sales, "Sales")


Employees:
                    missing_count  missing_%
salary                          5       4.07
performance_rating              5       4.07


In [14]:
# Show employees with missing salary
print("── Employees with missing salary ──")
employees[employees['salary'].isna()]

── Employees with missing salary ──


,employee_id,full_name,email,department,job_title,city,country,hire_date,salary,performance_rating,manager_id
4,1005,Christopher Nelson,christopher.nelson@techcorp.com,Finance,Controller,Dubai,UAE,2022-06-29,NaN,3.2,1008
5,1006,Joseph Kelly,joseph.kelly@techcorp.com,Sales,Business Development Rep,Amsterdam,Netherlands,2022-02-19,NaN,3.4,1001
17,1018,carol rogers,carol.rogers@techcorp.com,HR,HR Manager,Dubai,UAE,2021-05-07,NaN,3.3,1003
38,1039,Ryan Hall,ryan.hall@techcorp.com,HR,HR Generalist,Paris,France,2018-04-07,NaN,2.6,1003
63,1064,Jonathan Price,jonathan.price@techcorp.com,Customer Support,Customer Success Manager,Tokyo,Japan,2022-11-02,NaN,3.7,1012


### Handling Strategy for `employees.salary`

| Option | Method | When to use |
|--------|--------|-------------|
| 1 | `dropna()` — remove affected rows | When missing data is random and rows are expendable |
| 2 | Fixed fill — e.g. `fillna(0)` | When zero / a known default is meaningful |
| 3 | Statistical imputation — mean / median / mode | When you want to preserve distribution (our choice here) |

In [15]:
employees_cleaned = employees.dropna()

In [16]:
employees_cleaned

,employee_id,full_name,email,department,job_title,city,country,hire_date,salary,performance_rating,manager_id
0,1001,Gregory White,gregory.white@techcorp.com,Sales,Account Manager,Mumbai,India,2021-02-06,119704.35,4.9,1006
1,1002,Christine Sanders,christine.sanders@techcorp.com,HR,HR Generalist,Amsterdam,Netherlands,2022-06-10,54396.61,3.6,1003
2,1003,Donald Wright,donald.wright@techcorp.com,HR,Recruiter,Cape Town,South Africa,2018-11-27,55221.15,1.9,1002
3,1004,Richard Edwards,richard.edwards@techcorp.com,Engineering,QA Engineer,São Paulo,Brazil,2018-03-04,119468.82,2.6,1007
6,1007,Betty Jackson,betty.jackson@techcorp.com,Engineering,Software Engineer,Paris,France,2023-05-01,107240.14,4.1,1004
...,...,...,...,...,...,...,...,...,...,...,...
117,1118,Gary Lopez,gary.lopez@techcorp.com,Finance,Finance Manager,Toronto,Canada,2021-09-27,75316.41,1.4,1008
119,1120,Cynthia Kelly,cynthia.kelly@techcorp.com,Marketing,Digital Marketing Manager,Singapore,Singapore,2018-04-28,94373.72,2.3,1013
120,1011,Jason Moore,jason.moore@techcorp.com,Finance,Financial Analyst,San Francisco,USA,2020-12-01,81701.43,1.8,1008
121,1026,Jacob Nguyen,jacob.nguyen@techcorp.com,Engineering,Senior Software Engineer,Dubai,UAE,2021-03-04,117328.63,2.8,1004


In [17]:
print(f"\nOriginal employees count: {len(employees)}")
print(f"Cleaned employees count: {len(employees_cleaned)}")


Original employees count: 123
Cleaned employees count: 113


In [19]:
print(employees['salary'].describe())

count       118.000000
mean      98141.225763
std       32358.000110
min       40186.040000
25%       69177.350000
50%      102298.750000
75%      124714.742500
max      149829.990000
Name: salary, dtype: float64


In [20]:
print(employees.groupby('department')['salary'].median())

department
Customer Support     98764.275
Engineering         107899.335
Finance             105906.865
HR                   86249.100
Marketing           100777.760
Sales                88808.850
Name: salary, dtype: float64


In [21]:
print(employees['salary'].median())

102298.75


In [22]:
employees['salary'] = employees.groupby('department')['salary'].transform(lambda x: x.fillna(x.median()))

In [23]:
employees.isna().sum()

employee_id           0
full_name             0
email                 0
department            0
job_title             0
city                  0
country               0
hire_date             0
salary                0
performance_rating    5
manager_id            0
dtype: int64

In [25]:
overall_mean = employees['performance_rating'].mean().round(2)

In [26]:
overall_mean

np.float64(3.05)

In [27]:
employees['performance_rating'] = employees['performance_rating'].fillna(overall_mean)

In [28]:
employees.isna().sum()

employee_id           0
full_name             0
email                 0
department            0
job_title             0
city                  0
country               0
hire_date             0
salary                0
performance_rating    0
manager_id            0
dtype: int64

In [29]:
prod[prod['price'].isna()][['product_id', 'product_name', 'category', 'price']]

,product_id,product_name,category,price
7,2008,NexPhone Ultra,Phone,NaN
26,2027,CloudSync Pro,Software,NaN
33,2034,AutoReport,Software,NaN


In [30]:
print(prod.groupby('category')['price'].median())

category
Accessories      82.855
Laptop         1504.810
Phone           631.985
Software        137.260
Tablet          572.830
Name: price, dtype: float64


In [31]:
prod['price'] = prod.groupby('category')['price'].transform(lambda x: x.fillna(x.median()))

In [32]:
prod.isna().sum()

product_id      0
product_name    0
category        0
price           0
launch_date     0
dtype: int64

In [33]:
prod

,product_id,product_name,category,price,launch_date
0,2001,ProBook Elite,Laptop,1504.810,2021-05-08
1,2002,UltraSlim 14,Laptop,985.430,2023-01-28
2,2003,WorkStation X,Laptop,1585.910,2021-04-29
3,2004,DevPad Pro,Laptop,1387.220,2023-09-04
4,2005,NanoBook Air,Laptop,1926.860,2021-04-15
5,2006,SmartX One,Phone,426.170,2020-07-26
6,2007,ConnectPro 12,Phone,573.820,2021-01-28
7,2008,NexPhone Ultra,Phone,631.985,2021-06-29
8,2009,VoiceEdge S,Phone,690.150,2023-10-18
9,2010,PixelMax 5G,Phone,693.870,2022-11-05


In [34]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 410 entries, 0 to 409
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   sale_id        410 non-null    int64  
 1   date           410 non-null    object 
 2   employee_id    410 non-null    int64  
 3   product_id     410 non-null    int64  
 4   region         410 non-null    object 
 5   units_sold     410 non-null    int64  
 6   unit_price     410 non-null    float64
 7   discount       405 non-null    float64
 8   total_revenue  410 non-null    float64
dtypes: float64(3), int64(4), object(2)
memory usage: 29.0+ KB


In [35]:
sales['discount'] = sales['discount'].fillna(0)

In [36]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 410 entries, 0 to 409
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   sale_id        410 non-null    int64  
 1   date           410 non-null    object 
 2   employee_id    410 non-null    int64  
 3   product_id     410 non-null    int64  
 4   region         410 non-null    object 
 5   units_sold     410 non-null    int64  
 6   unit_price     410 non-null    float64
 7   discount       410 non-null    float64
 8   total_revenue  410 non-null    float64
dtypes: float64(3), int64(4), object(2)
memory usage: 29.0+ KB


In [38]:
sales[sales['sale_id'] == 3328]

,sale_id,date,employee_id,product_id,region,units_sold,unit_price,discount,total_revenue
327,3328,2023-11-28,1008,2002,Europe,19,1025.77,0.0,18710.04


In [39]:
for name, df in sheets.items():
    print(f"\n {'='*40}")
    print(f"    Sheet: {name}")
    print(f"{'='*40}")
    missing = df.isna().sum()
    print(missing[missing > 0] if missing.any() else "No missing values") 


    Sheet: Employees
No missing values

    Sheet: Departments
No missing values

    Sheet: Products
No missing values

    Sheet: Sales
No missing values


### ✅ Phase 3 Summary — Missing Value Fixes Applied

| Sheet | Column | Strategy |
|-------|--------|----------|
| Employees | `salary` | Filled with **median salary per department** |
| Employees | `performance_rating` | Filled with **overall mean rating** |
| Products | `price` | Filled with **median price per category** |
| Sales | `discount` | Filled with **0** (no discount assumed) |

---

## Phase 4 — Finding & Removing Duplicates

Earlier our shapes hinted at extra rows: Employees had 123 (expected 120) and Sales had 410 (expected 400).
Here we pinpoint and remove the duplicate records.

In [42]:
print(employees.shape)
print(dept.shape)
print(prod.shape)
print(sales.shape)

(123, 11)
(6, 4)
(35, 5)
(410, 9)


In [43]:
employees['employee_id'].nunique()

120

In [44]:
dept['department_id'].nunique()

6

In [45]:
prod['product_id'].nunique()

35

In [46]:
sales['sale_id'].nunique()

400

In [51]:
employees[employees.duplicated(subset='employee_id', keep='first')]

,employee_id,full_name,email,department,job_title,city,country,hire_date,salary,performance_rating,manager_id
120,1011,Jason Moore,jason.moore@techcorp.com,Finance,Financial Analyst,San Francisco,USA,2020-12-01,81701.43,1.8,1008
121,1026,Jacob Nguyen,jacob.nguyen@techcorp.com,Engineering,Senior Software Engineer,Dubai,UAE,2021-03-04,117328.63,2.8,1004
122,1051,Samantha King,samantha.king@techcorp.com,Sales,Business Development Rep,Cape Town,South Africa,2022-10-18,101699.76,1.4,1006


In [50]:
employees.duplicated().sum()

np.int64(3)